### ISSUE TITLE
[RFE] Please provide `%excludes` section

### ISSUE THREAD

**Issue by voxik** [rpm-software-management/rpm#1852]
There is `%exclude` macro which excludes file from package. However, it does not exclude the file from other processing such as generating debuginfo or what not and it causes confusion (there is probably more examples in #994).

So I was thinking wouldn't it be possible to have `%excludes` section instead? The idea is that the `%excludes` section would list files similarly to the `%files` sections, but these files would not be packaged. Let me give a brief example:

~~~
... snip ...

%install

... snip ...

mkdir -p %{buildroot}%{_sysconfdir}/libreport/events.d/
cp -a .%{gem_instdir}/config/ruby_event.conf %{buildroot}%{_sysconfdir}/libreport/events.d/

... snip ...

%files
%{gem_instdir}
%exclude %{gem_instdir}/config
%config(noreplace) %{_sysconfdir}/libreport/events.d/ruby_event.conf

%excludes
%{gem_instdir}/config

... snip ...
~~~

Note that the `%{gem_instdir}/config` is excluded form the main package, but then it is not obvious if it should not be included elsewhere. Therefore it is listed for the second time in the `%excludes` section to make it clear it is not forgotten and it should really be excluded.

In the simple scenario, I can imagine the functionality as RPM creating some `%{name}-exclude` subpackage containing such files and later just `rm %{name}-exclude` such package. Of course, it could also mean that some processing of such files is ignored at some point but that is just optimization.

**Comment by voxik** [comment]
Actually, the example given in https://github.com/rpm-software-management/rpm/issues/994#issuecomment-570681654 cold look like:

~~~
... snip ...

%files -n python3-%{pypi_name}
%license LICENSE
%doc README.rst
%{python3_sitelib}/%{pypi_name}/
%{python3_sitelib}/%{pypi_name}-%{version}.dist-info/

%excludes
%{python3_sitelib}/%{pypi_name}/tests

... snip ...
~~~

**Comment by rguenth** [comment]
I've opened kind-of a duplicate with https://github.com/rpm-software-management/rpm/issues/2555 since I have
the need to conditionally "drop" built subpackages.  So my idea is to add the ability to do this by turning
an existing %files into what you describe as %excludes section with a new option to %files (which I can then add
conditionally).  Implementation wise (without any of your suggested optimization) it could simply skip the
final .rpm file generation for such subpackages either by turning the name of the subpackage to something special
or by adding a flag to the metadata tracked.

**Comment by dmnks** [comment]
Yup, people seem to be interested in having native support for excludes. What remains to be seen is whether want package-level excludes (i.e. a preamble tag such as `Exclude: pkg1 pkg2 ...` as discussed in #2555) or file-level ones (i.e. an `%excludes` section as proposed here).

> Note that the `%{gem_instdir}/config` is excluded form the main package, but then it is not obvious if it should not be included elsewhere. Therefore it is listed for the second time in the `%excludes` section to make it clear it is not forgotten and it should really be excluded.

This seems redundant, though. Why would you need an `%exclude` in the `%files` section at all in this case?

**Comment by voxik** [comment]
> > Note that the `%{gem_instdir}/config` is excluded form the main package, but then it is not obvious if it should not be included elsewhere. Therefore it is listed for the second time in the `%excludes` section to make it clear it is not forgotten and it should really be excluded.
> 
> This seems redundant, though. Why would you need an `%exclude` in the `%files` section at all in this case?

Please note that in this specific case, the whole `%{gem_instdir}` is included in the package, except the `%{gem_instdir}/config` entry. However, it does not make clear in global context if `%{gem_instdir}/config` should be included elsewhere. To make it clear, it would need to be listed in other package or in `%excludes` section, otherwise it would be reported as an error (similarly if you miss some file).

IOW, this is to  prevent mistakes.

**Comment by dmnks** [comment]
Oh! My bad, I didn't realize this. Of course, this is a good point and something to consider if/when we design the solution.

I still don't like the duplicity, though. We could perhaps make it so that the `%excludes` section would just be "syntactic sugar" for explicit `%exclude` lines in all `%files` sections (plus the fact that RPM wouldn't complain about unpackaged files at the end), but then, it might be harder to follow when reading the SPEC...

**Comment by voxik** [comment]
> I still don't like the duplicity, though

Neither do I. I just hope that my example is exceptional and the entry would be enough to list in `%excludes` section in most of the cases 😇



> Example of what I mean:

Huh, I think my brain is going to explode now thinking about the example :exploding_head: But does it mean that this would include the file after all?

~~~
%files
%{gem_instdir}/config

%excludes
%{gem_instdir}/config
~~~

Also, I believe that your examples are not really equivalent. Because one of the issues is that currently `%excludes` does exclude such item from package, but does not exclude it from other processing (which is actually one of the problems). This might also disqualify #2555

**Comment by dmnks** [comment]
> ```
> %files
> %{gem_instdir}/config
> 
> %excludes
> %{gem_instdir}/config
> ```
> 
> Also, I believe that your examples are not really equivalent. Because one of the issues is that currently `%excludes` does exclude such item from package, but does not exclude it from other processing (which is actually one of the problems). This might also disqualify #2555

Yes, as I said, this is all assuming that we would *not* complain about unpackaged files in the case of `%excludes`.
